xcm/xcm-executor/integration-tests/src/lib.rs

```diff
@@ -71,6 +72,62 @@ fn basic_buy_fees_message_executes() {
 	});
 }
 
+#[test]
+fn transact_recursion_limit_works() {
+	sp_tracing::try_init_simple();
+	let mut client = TestClientBuilder::new()
+		.set_execution_strategy(ExecutionStrategy::AlwaysWasm)
+		.build();
+
+	let mut msg = Xcm(vec![ClearOrigin]);
+	let mut max_weight = <XcmConfig as xcm_executor::Config>::Weigher::weight(&mut msg).unwrap();
+	let mut call = polkadot_test_runtime::RuntimeCall::Xcm(
+		pallet_xcm::Call::execute { message: Box::new(VersionedXcm::from(msg)), max_weight },
+	);
+
+	for _ in 0..10 {
+		msg = Xcm(vec![
+			WithdrawAsset((Parent, 1_000_000_000).into()),
+			BuyExecution { fees: (Parent, 1_000_000_000).into(), weight_limit: Unlimited },
+			Transact {
+				origin_kind: OriginKind::Xcm,
+				require_weight_at_most: call.get_dispatch_info().weight,
+				call: call.encode().into(),
+			},
+		]);
+		max_weight = <XcmConfig as xcm_executor::Config>::Weigher::weight(&mut msg).unwrap();
+		call = polkadot_test_runtime::RuntimeCall::Xcm(
+			pallet_xcm::Call::execute { message: Box::new(VersionedXcm::from(msg)), max_weight }
+		);
+	}
+
+	let mut block_builder = client.init_polkadot_block_builder();
+
+	let execute = construct_extrinsic(
+		&client,
+		call,
+		sp_keyring::Sr25519Keyring::Alice,
+		0,
+	);
+
+	block_builder.push_polkadot_extrinsic(execute).expect("pushes extrinsic");
+
+	let block = block_builder.build().expect("Finalizes the block").block;
+	let block_hash = block.hash();
+
+	futures::executor::block_on(client.import(sp_consensus::BlockOrigin::Own, block))
+		.expect("imports the block");
+
+	client.state_at(block_hash).expect("state should exist").inspect_state(|| {
+		assert!(polkadot_test_runtime::System::events().iter().any(|r| matches!(
+			r.event,
+			polkadot_test_runtime::RuntimeEvent::Xcm(pallet_xcm::Event::Attempted(
```

Comment:
This unit test is currently failing, and I couldn't figure out why. I tried a couple of `panic` statements and found out that somehow, the `RECURSION_COUNT` thread local variable has always stayed below the value of 2, so I'm wondering if the thread local from `environmental` is not working at all.